btf/core.go

```diff
@@ -1040,13 +1045,293 @@ func coreAreMembersCompatible(localType Type, targetType Type) error {
+ *     forward declaration of a struct or union, respectively, and enum vs.
+ *     enum64 (see below)
+ * Then, depending on type:
+ * - integers:
+ *   - match if size and signedness match
+ * - arrays & pointers:
+ *   - target types are recursively matched
+ * - structs & unions:
+ *   - local members need to exist in target with the same name
+ *   - for each member we recursively check match unless it is already behind a
+ *     pointer, in which case we only check matching names and compatible kind
+ * - enums:
+ *   - local variants have to have a match in target by symbolic name (but not
+ *     numeric value)
+ *   - size has to match (but enum may match enum64 and vice versa)
+ * - function pointers:
+ *   - number and position of arguments in local type has to match target
+ *   - for each argument and the return value we recursively check match
+ */
+func coreTypesMatch(localType Type, targetType Type, behindPtr bool, level int) error {
+	if level <= 0 {
+		return fmt.Errorf("types are nested too deep")
+	}
+
+	var (
+		localTs, targetTs typeDeque
+		l, t              = &localType, &targetType
+		depth             = 0
+	)
+
+	for ; l != nil && t != nil; l, t = localTs.Shift(), targetTs.Shift() {
+		if depth >= maxResolveDepth {
+			return errors.New("types are nested too deep")
+		}
+
+		localType = UnderlyingType(*l)
+		targetType = UnderlyingType(*t)
+
+		if !coreNamesMatch(localType.TypeName(), targetType.TypeName()) {
+			return fmt.Errorf("type name %q don't match %q: %w", localType.TypeName(), targetType.TypeName(), errIncompatibleTypes)
+		}
+
+		switch lv := (localType).(type) {
+		case *Void:
+			if _, ok := targetType.(*Void); !ok {
+				return fmt.Errorf("type mismatch between %v and %v: %w", localType, targetType, errIncompatibleTypes)
+			}
+
+		case *Fwd:
+			if behindPtr {
+				if tv, ok := targetType.(*Fwd); ok {
+					if lv.Kind != tv.Kind {
+						return fmt.Errorf("fwd kind mismatch between %v and %v: %w", localType, targetType, errIncompatibleTypes)
+					}
+
+					continue
+				}
+
+				if _, ok := targetType.(*Struct); ok && lv.Kind == FwdStruct {
+					continue
+				}
+
+				if _, ok := targetType.(*Union); ok && lv.Kind == FwdUnion {
+					continue
+				}
+
+				return fmt.Errorf("fwd kind mismatch between %v and %v: %w", localType, targetType, errIncompatibleTypes)
+			}
+
+			if tv, ok := targetType.(*Fwd); ok && tv.Kind == lv.Kind {
+				continue
+			}
+
+			return fmt.Errorf("fwd kind mismatch between %v and %v: %w", localType, targetType, errIncompatibleTypes)
+
+		case *Enum:
+			tv, ok := targetType.(*Enum)
+			if !ok {
+				return fmt.Errorf("type mismatch between %v and %v: %w", localType, targetType, errIncompatibleTypes)
+			}
+
+			if err := coreEnumsMatch(lv, tv); err != nil {
+				return err
+			}
+
+		case *Struct, *Union:
+			if behindPtr {
+				if reflect.TypeOf(localType) == reflect.TypeOf(targetType) {
+					continue
```

Comment:
I think returning just the error is alright. This keeps in style with other functions in this package where we do:
* 1 = nil
* 0 = errIncompatibleTypes (with details)
* a unexpected error condition (instead of panic)

You are right though that we should not `continue` here but ignore the queues and return like libbpf does, we should only continue where libbpf does `goto recur` if we want to match bahavior.